File: ai-engine/src/intelligence/groupPreferenceFusion.js
import { calculateConfidence } from "./confidenceScoring.js";
import { clamp, normalizeText, tokenize } from "../retrieval/textUtils.js";

export const defaultGroupFusionWeights = {
  interest: 1,
  mood: 0.85,
  mustHave: 1.35,
  avoid: 1.15,
  explicitPreference: 1.25
};

export function buildTripDNA(input = {}) {
  return fuseGroupPreferences(input);
}

export function fuseGroupPreferences({
  groupMembers = [],
  explicitPreferences = {},
  tripContext = {},
  weights = defaultGroupFusionWeights
} = {}) {
  const memberProfiles = groupMembers.map(normalizeMemberProfile);
  const explicitProfile = normalizeMemberProfile({
    id: "trip_preferences",
    name: "Trip preferences",
    ...explicitPreferences
  }, {
    source: "explicit"
  });
  const profilesForSignals = explicitProfile.hasSignal
    ? [...memberProfiles, explicitProfile]
    : memberProfiles;
  const memberCount = Math.max(memberProfiles.length, 1);
  const signalCounts = collectSignalCounts(profilesForSignals, weights);
  const consensusTerms = rankTerms(signalCounts.positive)
    .filter((term) => term.supporterCount >= consensusThreshold(memberCount) || term.source === "explicit")
    .slice(0, 8);
  const fallbackTerms = rankTerms(signalCounts.positive).slice(0, 5);
  const softPreferences = (consensusTerms.length ? consensusTerms : fallbackTerms)
    .map((term) => term.term);
  const avoidTerms = rankTerms(signalCounts.avoid).slice(0, 8);
  const budgetProfile = buildBudgetProfile(memberProfiles, tripContext);
  const paceProfile = buildPaceProfile(memberProfiles, explicitPreferences);
  const fairness = buildFairnessProfile(memberProfiles, signalCounts.positive, memberCount);
  const conflicts = detectGroupConflicts({
    positiveTerms: rankTerms(signalCounts.positive),
    avoidTerms,
    budgetProfile,
    paceProfile,
    memberProfiles,
    tripContext
  });
  const hardConstraints = buildHardConstraints({
    avoidTerms,
    budgetProfile,
    memberProfiles,
    tripContext
  });
  const conflictLevel = conflicts.length >= 3
    ? "high"
    : conflicts.length >= 1 || fairness.minorityPreferences.length >= 3
      ? "medium"
      : "low";
  const confidenceReport = calculateConfidence({
    base: 0.88,
    missingInfo: memberProfiles.length ? [] : ["group members"],
    conflicts,
    groupConflictLevel: conflictLevel,
    outputCompleteness: softPreferences.length ? 1 : 0.55
  });
  const groupVibe = buildGroupVibe({ softPreferences, paceProfile, tripContext });

  return {
    stage: "group-preference-fusion-v1",
    generatedAt: new Date().toISOString(),
    inputSummary: {
      memberCount: memberProfiles.length,
      destination: tripContext.destination || null,
      days: tripContext.days || null,
      month: tripContext.month || null
    },
    tripDNA: {
      groupVibe,
      softPreferences,
      hardConstraints,
      dominantPace: paceProfile.dominantPace,
      budgetProfile,
      consensus: {
        terms: softPreferences,
        support: (consensusTerms.length ? consensusTerms : fallbackTerms).map(publicTermReport)
      },
      conflictLevel
    },
    memberProfiles: memberProfiles.map(publicMemberProfile),
    fairness,
    conflicts,
    recommendationHints: buildRecommendationHints({
      softPreferences,
      groupVibe,
      paceProfile,
      budgetProfile,
      tripContext,
      memberProfiles
    }),
    ...confidenceReport
  };
}

export function scorePlaceAgainstTripDNA(place = {}, fusion = {}) {
  const tripDNA = fusion.tripDNA || fusion;
  const softPreferences = tripDNA.softPreferences || [];
  const hardAvoids = (tripDNA.hardConstraints || [])
    .filter((constraint) => constraint.type === "avoid")
    .flatMap((constraint) => constraint.values || []);
  const placeTerms = new Set(tokenize([
    place.name,
    place.category,
    place.role,
    place.summary,
    place.semantic_summary,
    ...(place.mood || []),
    ...(place.tags || []),
    ...(place.ideal_for || place.bestFor || []),
    ...(place.retrieval_terms || place.retrievalTerms || [])
  ].join(" ")));
  const matchedPreferences = softPreferences
    .filter((preference) => tokenize(preference).some((term) => placeTerms.has(term)));
  const avoidedMatches = hardAvoids
    .filter((avoid) => tokenize(avoid).some((term) => placeTerms.has(term)));
  const consensusFit = softPreferences.length
    ? matchedPreferences.length / softPreferences.length
    : 0.55;
  const avoidPenalty = avoidedMatches.length * 0.18;
  const fairnessBoost = (fusion.fairness?.representationTargets || [])
    .some((target) => tokenize(target.preference).some((term) => placeTerms.has(term)))
    ? 0.08
    : 0;
  const score = clamp((consensusFit + fairnessBoost - avoidPenalty) * 100, 0, 100);

  return {
    placeId: place.id || null,
    name: place.name || null,
    score: Number(score.toFixed(2)),
    matchedPreferences,
    avoidedMatches,
    fairnessBoostApplied: fairnessBoost > 0
  };
}

function normalizeMemberProfile(member = {}, { source = "member" } = {}) {
  const interests = normalizeList(member.interests);
  const moods = normalizeList(member.moods);
  const mustHave = normalizeList(member.mustHave || member.must_haves || member.musts);
  const avoid = normalizeList(member.avoid || member.dislikes || member.blockedPreferences);
  const constraints = normalizeConstraints(member.constraints || {});
  const pace = normalizeText(member.pace || member.travelPace || member.travelStyle?.pace);
  const travelStyle = normalizeText(member.travelStyle?.primaryIntent || member.travelStyle || member.style);
  const id = member.id || member.name || source;

  return {
    id,
    name: member.name || id,
    source,
    interests,
    moods,
    mustHave,
    avoid,
    constraints,
    pace,
    travelStyle,
    budget: normalizeBudget(member.budget || member.maxBudget || member.budgetInr),
    hasSignal: [
      interests,
      moods,
      mustHave,
      avoid,
      Object.keys(constraints),
      pace ? [pace] : []
    ].some((values) => values.length)
  };
}

function normalizeList(value) {
  if (!value) return [];
  const values = Array.isArray(value) ? value : [value];
  return [...new Set(values
    .flatMap((item) => String(item || "").split(/[,/]/))
    .map((item) => normalizeText(item))
    .filter(Boolean))];
}

function normalizeConstraints(constraints = {}) {
  return Object.fromEntries(Object.entries(constraints)
    .filter(([, value]) => value !== undefined && value !== null && value !== false)
    .map(([key, value]) => [normalizeText(key), value === true ? true : normalizeText(value)]));
}

function normalizeBudget(value) {
  const budget = Number(value);
  return Number.isFinite(budget) && budget > 0 ? budget : null;
}

function collectSignalCounts(profiles, weights) {
  const positive = new Map();
  const avoid = new Map();

  for (const profile of profiles) {
    addTerms(positive, profile.interests, profile, weights.interest);
    addTerms(positive, profile.moods, profile, weights.mood);
    addTerms(positive, profile.mustHave, profile, weights.mustHave);
    if (profile.travelStyle) addTerms(positive, [profile.travelStyle], profile, weights.interest);
    addTerms(avoid, profile.avoid, profile, weights.avoid);

    if (profile.source === "explicit") {
      for (const report of positive.values()) {
        if (report.memberIds.has(profile.id)) report.score *= weights.explicitPreference;
      }
    }
  }

  return { positive, avoid };
}

function addTerms(target, terms, profile, weight) {
  for (const term of terms) {
    const current = target.get(term) || {
      term,
      score: 0,
      memberIds: new Set(),
      names: new Set(),
      source: profile.source
    };
    current.score += weight;
    current.memberIds.add(profile.id);
    current.names.add(profile.name);
    if (profile.source === "explicit") current.source = "explicit";
    target.set(term, current);
  }
}

function rankTerms(signalMap) {
  return [...signalMap.values()]
    .map((term) => ({
      ...term,
      supporterCount: term.memberIds.size,
      members: [...term.names],
      score: Number(term.score.toFixed(2))
    }))
    .sort((a, b) => b.supporterCount - a.supporterCount || b.score - a.score || a.term.localeCompare(b.term));
}

function consensusThreshold(memberCount) {
  if (memberCount <= 1) return 1;
  if (memberCount === 2) return 1;
  return Math.ceil(memberCount / 2);
}

function publicTermReport(term) {
  return {
    term: term.term,
    supporterCount: term.supporterCount,
    members: term.members,
    score: term.score
  };
}

function buildBudgetProfile(memberProfiles, tripContext = {}) {
  const memberBudgets = memberProfiles.map((member) => member.budget).filter(Boolean);
  const tripBudget = normalizeBudget(tripContext.budget);
  const budgets = [...memberBudgets, tripBudget].filter(Boolean).sort((a, b) => a - b);
  if (!budgets.length) {
    return {
      status: "unknown",
      min: null,
      max: null,
      median: null,
      sharedCap: null,
      spreadRatio: null
    };
  }

  const median = budgets[Math.floor(budgets.length / 2)];
  const min = budgets[0];
  const max = budgets[budgets.length - 1];

  return {
    status: max / min >= 2 ? "spread" : "aligned",
    min,
    max,
    median,
    sharedCap: min,
    spreadRatio: Number((max / min).toFixed(2))
  };
}

function buildPaceProfile(memberProfiles, explicitPreferences = {}) {
  const paceCounts = new Map();
  const explicitPace = normalizeText(explicitPreferences.pace || explicitPreferences.travelPace);

  for (const pace of [...memberProfiles.map((member) => member.pace), explicitPace].filter(Boolean)) {
    paceCounts.set(pace, (paceCounts.get(pace) || 0) + 1);
  }

  const ranked = [...paceCounts.entries()].sort((a, b) => b[1] - a[1]);
  return {
    dominantPace: ranked[0]?.[0] || "balanced",
    paceCounts: Object.fromEntries(ranked),
    hasConflict: ranked.length >= 2 && ranked[0][1] === ranked[1][1]
  };
}

function buildFairnessProfile(memberProfiles, positiveSignals, memberCount) {
  const rankedTerms = rankTerms(positiveSignals);
  const minorityPreferences = rankedTerms
    .filter((term) => term.supporterCount === 1 && memberCount >= 3)
    .map(publicTermReport);
  const representationTargets = minorityPreferences
    .slice(0, 4)
    .map((term) => ({
      preference: term.term,
      member: term.members[0],
      reason: "Keep at least one visible win for a minority preference."
    }));
  const underrepresentedMembers = memberProfiles
    .filter((member) => minorityPreferences.some((term) => term.members.includes(member.name)))
    .map((member) => member.name);

  return {
    minorityPreferences,
    representationTargets,
    underrepresentedMembers: [...new Set(underrepresentedMembers)],
    fairnessRule: memberCount >= 3
      ? "Do not average minority travelers out of the plan."
      : "Small-group plan can optimize for direct consensus."
  };
}

function detectGroupConflicts({
  positiveTerms,
  avoidTerms,
  budgetProfile,
  paceProfile,
  memberProfiles,
  tripContext
}) {
  const terms = new Set(positiveTerms.map((term) => term.term));
  const avoids = new Set(avoidTerms.map((term) => term.term));
  const conflicts = [];

  addOppositionConflict(conflicts, terms, ["quiet", "peaceful", "slow", "calm"], ["nightlife", "party", "energetic", "social"], {
    type: "quiet-nightlife",
    message: "The group wants quiet pacing and high-energy social blocks.",
    tradeoff: "Use a quiet base with optional evening zones instead of making nightlife the default."
  });
  addOppositionConflict(conflicts, terms, ["adventure", "trek", "hiking", "wild"], ["low fatigue", "accessible", "parents", "comfort"], {
    type: "adventure-fatigue",
    message: "Adventure interest is competing with low-fatigue or comfort needs.",
    tradeoff: "Prefer scenic low-risk adventure and cap long walking blocks."
  });
  addOppositionConflict(conflicts, terms, ["budget", "cheap", "low cost"], ["luxury", "premium", "comfort"], {
    type: "budget-comfort",
    message: "Budget and premium comfort expectations are both present.",
    tradeoff: "Keep the base plan budget-safe and mark premium upgrades separately."
  });

  if (budgetProfile.status === "spread") {
    conflicts.push({
      type: "budget-spread",
      message: `Group budgets range from ${budgetProfile.min} to ${budgetProfile.max}.`,
      tradeoff: "Rank against the lowest shared cap, then show paid upgrades as opt-ins."
    });
  }
  if (paceProfile.hasConflict) {
    conflicts.push({
      type: "pace-split",
      message: "No single pace dominates the group.",
      tradeoff: "Alternate anchor days with lighter recovery blocks."
    });
  }
  if (avoids.has("crowds") && terms.has("nightlife")) {
    conflicts.push({
      type: "crowd-nightlife",
      message: "Nightlife usually raises crowd exposure.",
      tradeoff: "Pick controlled evening venues over dense party strips."
    });
  }
  if (tripContext.days && memberProfiles.length >= 4 && Number(tripContext.days) <= 2) {
    conflicts.push({
      type: "short-trip-large-group",
      message: "A short trip with a larger group leaves little room for everyone to get a win.",
      tradeoff: "Prioritize two shared anchors and one optional split activity."
    });
  }

  return conflicts;
}

function addOppositionConflict(conflicts, terms, left, right, report) {
  if (left.some((term) => terms.has(term)) && right.some((term) => terms.has(term))) {
    conflicts.push(report);
  }
}

function buildHardConstraints({ avoidTerms, budgetProfile, memberProfiles, tripContext }) {
  const constraints = [];
  if (budgetProfile.sharedCap) {
    constraints.push({
      type: "budget",
      value: budgetProfile.sharedCap,
      source: "lowest shared budget"
    });
  } else if (tripContext.budget) {
    constraints.push({
      type: "budget",
      value: tripContext.budget,
      source: "trip context"
    });
  }
  if (avoidTerms.length) {
    constraints.push({
      type: "avoid",
      values: avoidTerms.map((term) => term.term),
      source: "traveler dislikes"
    });
  }

  const constraintTerms = memberProfiles.flatMap((member) => Object.entries(member.constraints)
    .map(([key, value]) => value === true ? key : `${key}: ${value}`));
  if (constraintTerms.length) {
    constraints.push({
      type: "traveler-constraint",
      values: [...new Set(constraintTerms)],
      source: "member constraints"
    });
  }

  return constraints;
}

function buildGroupVibe({ softPreferences, paceProfile, tripContext }) {
  const pace = paceProfile.dominantPace || "balanced";
  const theme = softPreferences.slice(0, 2).join(" + ") || normalizeText(tripContext.groupType) || "shared";
  return `${pace} ${theme} trip`;
}

function buildRecommendationHints({
  softPreferences,
  groupVibe,
  paceProfile,
  budgetProfile,
  tripContext,
  memberProfiles
}) {
  const familySignals = memberProfiles.some((member) => (
    member.constraints.parents || member.interests.includes("family") || member.mustHave.includes("family")
  ));
  const groupType = normalizeText(tripContext.groupType)
    || (familySignals ? "family" : memberProfiles.length >= 3 ? "friends" : "solo");

  return {
    groupType,
    groupVibe,
    userPreferences: {
      interests: softPreferences,
      moods: softPreferences.filter((term) => ["slow", "quiet", "social", "romantic", "adventurous"].includes(term))
    },
    travelStyle: {
      pace: paceProfile.dominantPace,
      primaryIntent: softPreferences[0] || null
    },
    budgetConstraints: {
      maxBudget: budgetProfile.sharedCap,
      preferredBudget: budgetProfile.median
    },
    rankingNotes: [
      "Use consensus terms for the main route.",
      "Reserve at least one visible win for minority preferences.",
      "Treat hard constraints as filters, not soft scoring hints."
    ]
  };
}

function publicMemberProfile(member) {
  return {
    id: member.id,
    name: member.name,
    interests: member.interests,
    moods: member.moods,
    mustHave: member.mustHave,
    avoid: member.avoid,
    pace: member.pace || null,
    budget: member.budget
  };
}

File: ai-engine/app/engines/group_preference_fusion.py
from collections import Counter

from ..core.text import clamp, normalize_text, tokenize
from .confidence import calculate_confidence

DEFAULT_GROUP_FUSION_WEIGHTS = {
    "interest": 1.0,
    "mood": 0.85,
    "mustHave": 1.35,
    "avoid": 1.15,
    "explicitPreference": 1.25,
}


def build_trip_dna(payload: dict | None = None) -> dict:
    return fuse_group_preferences(**(payload or {}))


def fuse_group_preferences(
    *,
    groupMembers: list[dict] | None = None,
    group_members: list[dict] | None = None,
    explicitPreferences: dict | None = None,
    explicit_preferences: dict | None = None,
    tripContext: dict | None = None,
    trip_context: dict | None = None,
    weights: dict | None = None,
) -> dict:
    members = [normalize_member(member) for member in (groupMembers or group_members or [])]
    explicit = normalize_member(
        {"id": "trip_preferences", "name": "Trip preferences", **(explicitPreferences or explicit_preferences or {})},
        source="explicit",
    )
    active_weights = {**DEFAULT_GROUP_FUSION_WEIGHTS, **(weights or {})}
    profiles = [*members, explicit] if explicit["hasSignal"] else members
    positive, avoid = collect_signal_counts(profiles, active_weights)
    member_count = max(len(members), 1)
    ranked_positive = rank_terms(positive)
    consensus_terms = [
        term
        for term in ranked_positive
        if term["supporterCount"] >= consensus_threshold(member_count) or term["source"] == "explicit"
    ][:8]
    fallback_terms = ranked_positive[:5]
    soft_preferences = [term["term"] for term in (consensus_terms or fallback_terms)]
    avoid_terms = rank_terms(avoid)[:8]
    context = tripContext or trip_context or {}
    budget_profile = build_budget_profile(members, context)
    pace_profile = build_pace_profile(members, explicitPreferences or explicit_preferences or {})
    fairness = build_fairness_profile(members, positive, member_count)
    conflicts = detect_group_conflicts(
        positive_terms=ranked_positive,
        avoid_terms=avoid_terms,
        budget_profile=budget_profile,
        pace_profile=pace_profile,
        member_profiles=members,
        trip_context=context,
    )
    conflict_level = (
        "high" if len(conflicts) >= 3 else "medium" if conflicts or len(fairness["minorityPreferences"]) >= 3 else "low"
    )
    confidence = calculate_confidence(
        base=0.88,
        missing_info=[] if members else ["group members"],
        conflicts=conflicts,
        group_conflict_level=conflict_level,
        output_completeness=1 if soft_preferences else 0.55,
    )
    group_vibe = build_group_vibe(soft_preferences, pace_profile, context)
    trip_dna = {
        "groupVibe": group_vibe,
        "softPreferences": soft_preferences,
        "hardConstraints": build_hard_constraints(avoid_terms, budget_profile, members, context),
        "dominantPace": pace_profile["dominantPace"],
        "budgetProfile": budget_profile,
        "consensus": {
            "terms": soft_preferences,
            "support": [public_term(term) for term in (consensus_terms or fallback_terms)],
        },
        "conflictLevel": conflict_level,
    }

    return {
        "stage": "group-preference-fusion-v1",
        "inputSummary": {
            "memberCount": len(members),
            "destination": context.get("destination"),
            "days": context.get("days"),
            "month": context.get("month"),
        },
        "tripDNA": trip_dna,
        "memberProfiles": [public_member(member) for member in members],
        "fairness": fairness,
        "conflicts": conflicts,
        "recommendationHints": build_recommendation_hints(
            soft_preferences, group_vibe, pace_profile, budget_profile, context, members
        ),
        **confidence,
    }


def score_place_against_trip_dna(place: dict, fusion: dict) -> dict:
    trip_dna = fusion.get("tripDNA", fusion)
    soft_preferences = trip_dna.get("softPreferences", [])
    hard_avoids = [
        value
        for constraint in trip_dna.get("hardConstraints", [])
        if constraint.get("type") == "avoid"
        for value in constraint.get("values", [])
    ]
    place_terms = set(
        tokenize(
            " ".join(
                [
                    place.get("name", ""),
                    place.get("category", ""),
                    place.get("summary", ""),
                    place.get("semantic_summary", ""),
                    *place.get("mood", []),
                    *place.get("tags", []),
                    *place.get("ideal_for", place.get("bestFor", [])),
                    *place.get("retrieval_terms", place.get("retrievalTerms", [])),
                ]
            )
        )
    )
    matched = [pref for pref in soft_preferences if any(term in place_terms for term in tokenize(pref))]
    avoided = [avoid for avoid in hard_avoids if any(term in place_terms for term in tokenize(avoid))]
    consensus_fit = len(matched) / len(soft_preferences) if soft_preferences else 0.55
    fairness_boost = (
        0.08
        if any(
            any(term in place_terms for term in tokenize(target["preference"]))
            for target in fusion.get("fairness", {}).get("representationTargets", [])
        )
        else 0
    )
    score = clamp((consensus_fit + fairness_boost - len(avoided) * 0.18) * 100, 0, 100)
    return {
        "placeId": place.get("id"),
        "name": place.get("name"),
        "score": round(score, 2),
        "matchedPreferences": matched,
        "avoidedMatches": avoided,
        "fairnessBoostApplied": fairness_boost > 0,
    }


def normalize_member(member: dict, source: str = "member") -> dict:
    interests = normalize_list(member.get("interests"))
    moods = normalize_list(member.get("moods"))
    must_have = normalize_list(member.get("mustHave") or member.get("must_haves") or member.get("musts"))
    avoid = normalize_list(member.get("avoid") or member.get("dislikes") or member.get("blockedPreferences"))
    constraints = normalize_constraints(member.get("constraints") or {})
    travel_style = member.get("travelStyle") or {}
    if not isinstance(travel_style, dict):
        travel_style = {"primaryIntent": travel_style}
    pace = normalize_text(member.get("pace") or member.get("travelPace") or travel_style.get("pace"))
    style = normalize_text(travel_style.get("primaryIntent") or member.get("style"))
    member_id = member.get("id") or member.get("name") or source
    return {
        "id": member_id,
        "name": member.get("name") or member_id,
        "source": source,
        "interests": interests,
        "moods": moods,
        "mustHave": must_have,
        "avoid": avoid,
        "constraints": constraints,
        "pace": pace,
        "travelStyle": style,
        "budget": normalize_budget(member.get("budget") or member.get("maxBudget") or member.get("budgetInr")),
        "hasSignal": any([interests, moods, must_have, avoid, constraints, pace]),
    }


def normalize_list(value: object) -> list[str]:
    if not value:
        return []
    values = value if isinstance(value, list) else [value]
    normalized: list[str] = []
    for item in values:
        normalized.extend(normalize_text(part) for part in str(item).split(","))
    return sorted({item for item in normalized if item})


def normalize_constraints(constraints: dict) -> dict:
    return {
        normalize_text(key): True if value is True else normalize_text(value)
        for key, value in constraints.items()
        if value not in (None, False, "")
    }


def normalize_budget(value: object) -> int | None:
    try:
        budget = int(value)
    except (TypeError, ValueError):
        return None
    return budget if budget > 0 else None


def collect_signal_counts(profiles: list[dict], weights: dict) -> tuple[dict, dict]:
    positive: dict[str, dict] = {}
    avoid: dict[str, dict] = {}
    for profile in profiles:
        add_terms(positive, profile["interests"], profile, weights["interest"])
        add_terms(positive, profile["moods"], profile, weights["mood"])
        add_terms(positive, profile["mustHave"], profile, weights["mustHave"])
        if profile["travelStyle"]:
            add_terms(positive, [profile["travelStyle"]], profile, weights["interest"])
        add_terms(avoid, profile["avoid"], profile, weights["avoid"])
        if profile["source"] == "explicit":
            for report in positive.values():
                if profile["id"] in report["memberIds"]:
                    report["score"] *= weights["explicitPreference"]
    return positive, avoid


def add_terms(target: dict, terms: list[str], profile: dict, weight: float) -> None:
    for term in terms:
        report = target.setdefault(
            term,
            {"term": term, "score": 0.0, "memberIds": set(), "names": set(), "source": profile["source"]},
        )
        report["score"] += weight
        report["memberIds"].add(profile["id"])
        report["names"].add(profile["name"])
        if profile["source"] == "explicit":
            report["source"] = "explicit"


def rank_terms(signal_map: dict) -> list[dict]:
    terms = [
        {
            **report,
            "supporterCount": len(report["memberIds"]),
            "members": sorted(report["names"]),
            "score": round(report["score"], 2),
        }
        for report in signal_map.values()
    ]
    return sorted(terms, key=lambda term: (-term["supporterCount"], -term["score"], term["term"]))


def consensus_threshold(member_count: int) -> int:
    if member_count <= 2:
        return 1
    return (member_count + 1) // 2


def public_term(term: dict) -> dict:
    return {
        "term": term["term"],
        "supporterCount": term["supporterCount"],
        "members": term["members"],
        "score": term["score"],
    }


def build_budget_profile(members: list[dict], context: dict) -> dict:
    budgets = sorted(
        budget
        for budget in [*(member["budget"] for member in members), normalize_budget(context.get("budget"))]
        if budget
    )
    if not budgets:
        return {"status": "unknown", "min": None, "max": None, "median": None, "sharedCap": None, "spreadRatio": None}
    minimum, maximum = budgets[0], budgets[-1]
    return {
        "status": "spread" if maximum / minimum >= 2 else "aligned",
        "min": minimum,
        "max": maximum,
        "median": budgets[len(budgets) // 2],
        "sharedCap": minimum,
        "spreadRatio": round(maximum / minimum, 2),
    }


def build_pace_profile(members: list[dict], explicit: dict) -> dict:
    counts = Counter(member["pace"] for member in members if member["pace"])
    explicit_pace = normalize_text(explicit.get("pace") or explicit.get("travelPace"))
    if explicit_pace:
        counts[explicit_pace] += 1
    ranked = counts.most_common()
    return {
        "dominantPace": ranked[0][0] if ranked else "balanced",
        "paceCounts": dict(ranked),
        "hasConflict": len(ranked) >= 2 and ranked[0][1] == ranked[1][1],
    }


def build_fairness_profile(members: list[dict], positive: dict, member_count: int) -> dict:
    minority = [public_term(term) for term in rank_terms(positive) if term["supporterCount"] == 1 and member_count >= 3]
    targets = [
        {
            "preference": term["term"],
            "member": term["members"][0],
            "reason": "Keep at least one visible win for a minority preference.",
        }
        for term in minority[:4]
    ]
    underrepresented = sorted({target["member"] for target in targets})
    return {
        "minorityPreferences": minority,
        "representationTargets": targets,
        "underrepresentedMembers": underrepresented,
        "fairnessRule": "Do not average minority travelers out of the plan."
        if member_count >= 3
        else "Small-group plan can optimize for direct consensus.",
    }


def detect_group_conflicts(
    *,
    positive_terms: list[dict],
    avoid_terms: list[dict],
    budget_profile: dict,
    pace_profile: dict,
    member_profiles: list[dict],
    trip_context: dict,
) -> list[dict]:
    terms = {term["term"] for term in positive_terms}
    avoids = {term["term"] for term in avoid_terms}
    conflicts: list[dict] = []
    add_opposition(
        conflicts,
        terms,
        {"quiet", "peaceful", "slow", "calm"},
        {"nightlife", "party", "energetic", "social"},
        {
            "type": "quiet-nightlife",
            "message": "The group wants quiet pacing and high-energy social blocks.",
            "tradeoff": "Use a quiet base with optional evening zones instead of making nightlife the default.",
        },
    )
    add_opposition(
        conflicts,
        terms,
        {"adventure", "trek", "hiking", "wild"},
        {"low fatigue", "accessible", "parents", "comfort"},
        {
            "type": "adventure-fatigue",
            "message": "Adventure interest is competing with low-fatigue or comfort needs.",
            "tradeoff": "Prefer scenic low-risk adventure and cap long walking blocks.",
        },
    )
    add_opposition(
        conflicts,
        terms,
        {"budget", "cheap", "low cost"},
        {"luxury", "premium", "comfort"},
        {
            "type": "budget-comfort",
            "message": "Budget and premium comfort expectations are both present.",
            "tradeoff": "Keep the base plan budget-safe and mark premium upgrades separately.",
        },
    )
    if budget_profile["status"] == "spread":
        conflicts.append(
            {
                "type": "budget-spread",
                "message": f"Group budgets range from {budget_profile['min']} to {budget_profile['max']}.",
                "tradeoff": "Rank against the lowest shared cap, then show paid upgrades as opt-ins.",
            }
        )
    if pace_profile["hasConflict"]:
        conflicts.append(
            {
                "type": "pace-split",
                "message": "No single pace dominates the group.",
                "tradeoff": "Alternate anchor days with lighter recovery blocks.",
            }
        )
    if "crowds" in avoids and "nightlife" in terms:
        conflicts.append(
            {
                "type": "crowd-nightlife",
                "message": "Nightlife usually raises crowd exposure.",
                "tradeoff": "Pick controlled evening venues over dense party strips.",
            }
        )
    if trip_context.get("days") and len(member_profiles) >= 4 and int(trip_context["days"]) <= 2:
        conflicts.append(
            {
                "type": "short-trip-large-group",
                "message": "A short trip with a larger group leaves little room for everyone to get a win.",
                "tradeoff": "Prioritize two shared anchors and one optional split activity.",
            }
        )
    return conflicts


def add_opposition(conflicts: list[dict], terms: set[str], left: set[str], right: set[str], report: dict) -> None:
    if terms.intersection(left) and terms.intersection(right):
        conflicts.append(report)


def build_hard_constraints(
    avoid_terms: list[dict], budget_profile: dict, members: list[dict], context: dict
) -> list[dict]:
    constraints: list[dict] = []
    if budget_profile.get("sharedCap"):
        constraints.append({"type": "budget", "value": budget_profile["sharedCap"], "source": "lowest shared budget"})
    elif context.get("budget"):
        constraints.append({"type": "budget", "value": context["budget"], "source": "trip context"})
    if avoid_terms:
        constraints.append(
            {"type": "avoid", "values": [term["term"] for term in avoid_terms], "source": "traveler dislikes"}
        )
    member_constraints = [
        key if value is True else f"{key}: {value}"
        for member in members
        for key, value in member["constraints"].items()
    ]
    if member_constraints:
        constraints.append(
            {"type": "traveler-constraint", "values": sorted(set(member_constraints)), "source": "member constraints"}
        )
    return constraints


def build_group_vibe(soft_preferences: list[str], pace_profile: dict, context: dict) -> str:
    theme = " + ".join(soft_preferences[:2]) or normalize_text(context.get("groupType")) or "shared"
    return f"{pace_profile['dominantPace']} {theme} trip"


def build_recommendation_hints(
    soft_preferences: list[str],
    group_vibe: str,
    pace_profile: dict,
    budget_profile: dict,
    context: dict,
    members: list[dict],
) -> dict:
    family = any(member["constraints"].get("parents") or "family" in member["interests"] for member in members)
    group_type = normalize_text(context.get("groupType")) or (
        "family" if family else "friends" if len(members) >= 3 else "solo"
    )
    return {
        "groupType": group_type,
        "groupVibe": group_vibe,
        "userPreferences": {
            "interests": soft_preferences,
            "moods": [
                term for term in soft_preferences if term in {"slow", "quiet", "social", "romantic", "adventurous"}
            ],
        },
        "travelStyle": {
            "pace": pace_profile["dominantPace"],
            "primaryIntent": soft_preferences[0] if soft_preferences else None,
        },
        "budgetConstraints": {
            "maxBudget": budget_profile.get("sharedCap"),
            "preferredBudget": budget_profile.get("median"),
        },
        "rankingNotes": [
            "Use consensus terms for the main route.",
            "Reserve at least one visible win for minority preferences.",
            "Treat hard constraints as filters, not soft scoring hints.",
        ],
    }


def public_member(member: dict) -> dict:
    return {
        "id": member["id"],
        "name": member["name"],
        "interests": member["interests"],
        "moods": member["moods"],
        "mustHave": member["mustHave"],
        "avoid": member["avoid"],
        "pace": member["pace"] or None,
        "budget": member["budget"],
    }
